core/src/light_protocol/handler/sync/witnesses.rs

```diff
@@ -96,14 +102,54 @@ impl Witnesses {
     /// Get root hashes for `epoch` from local cache.
     #[inline]
     pub fn root_hashes_of(&self, epoch: u64) -> Option<(H256, H256, H256)> {
-        self.verified.read().get(&epoch).cloned()
+        let height = epoch + DEFERRED_STATE_EPOCH_COUNT;
+
+        if height > *self.latest_verified_header.read() {
+            return None;
+        }
+
+        match self.data_man.verified_blamed_roots_by_height(height) {
+            Some(roots) => Some(roots.into_tuple()),
+            None => {
+                // we set `latest_verified_header` before receiving the
+                // response for blamed headers. thus, in some cases, `None`
+                // might mean *haven't received yet* instead of *not blamed*.
+                if self.in_flight.read().contains(&height) {
+                    // TODO(thegaram): recover from this
+                    error!("Witness {} still in flight!", height);
+                    panic!("Witness requested still in flight!");
+                }
```

Comment:
An example for this triggered in a test:

```
2020-08-10T16:10:26.620575+02:00 DEBUG Witness Worker cfxcore::lig -
    Witness worker received: height = 1091, maybe_witness = Some(1092)

...

2020-08-10T16:10:27.446195+02:00 ERROR Socket IO Worker #0 cfxcore::lig -
    Witness 1091 still in flight!

...

2020-08-10T16:10:27.447615+02:00 DEBUG Socket IO Worker #1 cfxcore::lig -
    Validating witness info WitnessInfoWithHeight { height: 1092, state_root_hashes: [0x2ebe5b83e4be9a10c12d8d171eca52c5b3c9c970dd6428e820b9bf39dfb94cd4, 0x775a4de351b06fac2d750d2478a49383bce5877f7ade0d2ca11f41626b4d4cf4], receipt_hashes: [0x09f8709ea9f344a810811a373b30861568f5686e649d6177fd92ea2db7477508, 0x09f8709ea9f344a810811a373b30861568f5686e649d6177fd92ea2db7477508], bloom_hashes: [0xd397b3b043d87fcd6fad1291ff0bfd16401c274896d8c63a923727f077b8e0b5, 0xd397b3b043d87fcd6fad1291ff0bfd16401c274896d8c63a923727f077b8e0b5] }
```

Header 1091 is blamed by 1092 so we request 1092 as the witness. Another request (`GetBlooms`) needs the correct roots before we actually received them.

The best way to handle this is to signal the error to the RPC caller directly.